### PR TITLE
Remove test class exclusion for Abstract prefix and rename classes accordingly

### PR DESCRIPTION
**Pull request by rjernst**
While the list of having exclusions is small, it shouldn't be necessary
at all. Base test cases should be suffixed with TestCase so they are not
picked up by the test class name pattern. This same rule works for
abstract classes as well.

This change renames abstract tests to use the TestCase suffix, adds a
check in naming convention tests, and removes the exclusion from our
test runner configuration. It also excludes inner classes (the only exclude
we should have IMO), so that we have no need to @Ignore the inner test
classes for naming convention tests.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/NamingConventionTests.java [hunk 1]
```diff
@@ -48,11 +48,13 @@
 
     // see https://github.com/elasticsearch/elasticsearch/issues/9945
     public void testNamingConventions()
-            throws ClassNotFoundException, IOException, URISyntaxException {
+        throws ClassNotFoundException, IOException, URISyntaxException {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/NamingConventionTests.java]: This makes it harder to read but isn't really a problem with the commit per se.

File: core/src/test/java/org/elasticsearch/NamingConventionTests.java [hunk 2]
```diff
@@ -76,27 +78,33 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
...
-                                    if (isTestCase(clazz) == false) {
-                                        notImplementing.add(clazz);
-                                    }
-                                } else if (clazz.getName().endsWith("IT")) {
-                                    if (isTestCase(clazz) == false) {
-                                        notImplementing.add(clazz);
-                                    }
-                                    // otherwise fine
-                                } else if (isTestCase(clazz)) {
+                            //if (Modifier.isAbstract(clazz.getModifiers()) == false && Modifier.isInterface(clazz.getModifiers()) == false) {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/NamingConventionTests.java]: Did you intentionally leave this?

Comment (rjernst) [inline core/src/test/java/org/elasticsearch/NamingConventionTests.java]: No! Will remove.

File: core/src/test/java/org/elasticsearch/NamingConventionTests.java [hunk 3]
```diff
@@ -143,39 +151,57 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOEx
...
+        assertTrue("Found inner classes that are tests, which are excluded from the test runner:\n" + listClasses(innerClasses),
+            innerClasses.isEmpty());
+        assertTrue("Pure Unit-Test found must subclass one of [" + classesToSubclass +"]:\n" + listClasses(pureUnitTest),
+            pureUnitTest.isEmpty());
+        assertTrue("Classes ending with Test|Tests] must subclass [" + classesToSubclass +"]:\n" + listClasses(notImplementing),
+            notImplementing.isEmpty());
+        assertTrue("Subclasses of ESIntegTestCase should end with IT as they are integration tests:\n" + listClasses(integTestsInDisguise), integTestsInDisguise.isEmpty());
+    }
+
+    static String listClasses(Set<Class> classes) {
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/NamingConventionTests.java]: This is one of those cases where I prefer Guava's `Joiner.on("\n").join(classes)`

File: core/src/test/java/org/elasticsearch/NamingConventionTests.java (hunk 1 above)
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/NamingConventionTests.java]: was not intentional, will revert

**Comment by nik9000** [comment]
LGTM

File: core/src/test/java/org/elasticsearch/NamingConventionTests.java (hunk 3 above)
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/NamingConventionTests.java]: Sure, I'll change. That will make it easier to find places to switch to String.join in java 8.

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/NamingConventionTests.java]: There is both `String.join` and `StringJoiner` that were added in Java 8.

**Comment by dadoonet** [comment]
@rjernst This commit seems to break plugins like Azure:

```
Suite: org.elasticsearch.index.store.AbstractAzureFsTest
ERROR   0.01s J0 | AbstractAzureFsTest.initializationError <<<
   > Throwable #1: java.lang.RuntimeException: Suite class org.elasticsearch.index.store.AbstractAzureFsTest should be a concrete class (not abstract).
   > 	at com.carrotsearch.randomizedtesting.Validation$ClassValidation.isConcreteClass(Validation.java:90)
   > 	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
Completed [14/14] on J0 in 0.02s, 1 test, 1 error <<< FAILURES!


Tests with failures (first 3 out of 4):
  - org.elasticsearch.cloud.azure.AbstractAzureComputeServiceTest.initializationError
  - org.elasticsearch.cloud.azure.AbstractAzureRepositoryServiceTest.initializationError
  - org.elasticsearch.cloud.azure.AbstractAzureTest.initializationError
```

**Comment by rjernst** [comment]
I just pushed a fix for the plugins failures.

**Pull request merged**